File: sdp/model-free-LQR-design-by-Q-function-learning/automatica.py
import cvxpy as cp
import numpy as np
import scipy.linalg as linalg
import control



A = np.array([
    [0.2, 0.7, 0, 0],
    [0, 0.4, 0.7, 0],
    [0, 0, 0.5, 0.8],
    [0.7, 0, 0, 0.4]
])
B = np.eye(4)
controllability = np.linalg.matrix_rank(control.ctrb(A, B))
print(f'controllability matrix rank: {controllability}')
Q = 1*np.eye(4)
R = 1*np.eye(4)
Gamma = linalg.block_diag(Q, R)
n, m = B.shape
H = cp.Variable((n+m,n+m), hermitian=False, symmetric=True)
W = cp.Variable((n,n), hermitian=False, symmetric=True)

obj = cp.trace(W)
W_a = cp.bmat([[W, np.zeros((n,m))], [np.zeros((m, n+m))]])
cons = [H-W_a >> 0]

AB = cp.bmat([[A, B, np.zeros((n, m))]])
ZI = cp.bmat([[np.zeros((m, n+m)), np.eye(m)]])
ABI = cp.bmat([[AB], [ZI]])
GammaH = cp.bmat([[Gamma-H, np.zeros((m+n, m))], [np.zeros((m, 2*m+n))]])
cons += [ABI.T@H@ABI + GammaH >> 0 ]

prob = cp.Problem(cp.Maximize(obj), cons)
# 
print(cp.installed_solvers())
prob.solve(solver=cp.MOSEK, verbose=True)

H_opt = H.value
H22 = np.array(H_opt[n:, n:])
H12 = np.array(H_opt[0:n, n:])
print(f'H22: \n {H22} \n, H12: \n {H12}')
K_opt = -np.matmul(np.linalg.inv(H22), np.transpose(H12))
print(f'optimal cost: {prob.value}, \n W: \n {W.value}, \n H: \n {H.value}, \n K: \n {K_opt}')
# print(vars(prob))

File: sdp/model-free-LQR-design-by-Q-function-learning/automatica.m
clc;
clear;
close all;

%% System Matrix
A = [ 0.2 0.7 0 0;
    0 0.4 0.7 0;
    0 0 0.5 0.5;
    0.7 0 0 0.4];
B = eye(4);

Q = eye(4);
R = eye(4);

[n, m] = size(B);

Gamma = blkdiag(Q, R);


%% Optimization 
H = sdpvar(n+m, n+m);
W = sdpvar(n, n);
W_a = [W zeros(n,m); zeros(m, n+m)];

c_1 =  (H-W_a) >= 0;

ABI = [[A, B], zeros(n, m); zeros(n, n+m), eye(m)];
GammaH = [Gamma-H, zeros(m+n, m); zeros(m, 2*m+n)];
c_2 = (ABI'*H*ABI + GammaH) >= 0;


constraints = c_1 + c_2;

cost = -trace(W);


ops = sdpsettings('solver','mosek','verbose',0);
diagnostics = optimize(constraints, cost, ops)

W_opt = value(W)
H_opt = value(H)
K = -inv(H_opt(n+1:n+m, n+1:n+m))*H_opt(1:n, n+1:n+m)'


